### PR TITLE
Flexible AdapterRemoval

### PR DESCRIPTION
**Pull request by apeltzer**
This is based on @maxibor and @jfy133 s work and help and should provide the means to run:

- AdapterRemoval v2 on PE/SE data
- with or without read merging /collapsing
- with or without read trimming
- Skipping the entire module if no need for the module in general.

Thereby addressing #137, #64. 

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] If necessary, also make a PR on the [nf-core/eager branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/eager)
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Make sure your code lints (`nf-core lint .`).
 - [ ] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated

This might require some docs updating (open a PR against the `ar_optional` branch here) and some more testing, as I adjusted some stuff in my PR on top of @maxibor 's. 

Thanks A BUNCH for helping out on this both!

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -59,6 +60,8 @@ def helpMessage() {
       --clip_readlength             Specify read minimum length to be kept for downstream analysis
       --clip_min_read_quality       Specify minimum base quality for not trimming off bases
       --min_adap_overlap            Specify minimum adapter overlap
+      --skip_collapse               Skip merging Forward and Reverse reads together. (Only for PE samples)
```
Comment (phue) [inline main.nf]:
```suggestion
      --skip_collapse               Skip merging forward and reverse reads together. (Only for PE samples)
```

File: main.nf [hunk 2]
```diff
@@ -568,60 +580,85 @@ process fastp {
 /*
  * STEP 2 - Adapter Clipping / Read Merging
  */
-
-
+//Initialize empty channel if we skip adapterremoval entirely
+if(params.skip_adapterremoval) {
```
Comment (phue) [inline main.nf]:
```suggestion
if (params.skip_adapterremoval) {
```

File: main.nf [hunk 3]
```diff
@@ -568,60 +580,85 @@ process fastp {
...
 
     script:
-    prefix = reads[0].toString() - ~/(_R1)?(_trimmed)?(_val_1)?(\.fq)?(\.fastq)?(\.gz)?$/
-    //Readprefixing only required for PE data with merging
-    fixprefix = (params.singleEnd) ? "" : "AdapterRemovalFixPrefix ${prefix}.combined.fq.gz ${prefix}.combined.prefixed.fq.gz"
+    base = reads[0].baseName
     
-    if( !params.singleEnd ){
+    //PE, collapse & trim reads
+    if( !params.singleEnd && !params.skip_collapse && !params.skip_trim && !params.skip_adapterremoval){
```
Comment (phue) [inline main.nf]:
```suggestion
    if (!params.singleEnd && !params.skip_collapse && !params.skip_trim && !params.skip_adapterremoval){
```

Comment (phue) [reply main.nf]: whitespace

File: main.nf [hunk 4]
```diff
@@ -703,17 +755,31 @@ process circularmapper{
     
     script:
     filter = "${params.circularfilter}" ? '' : '-f true -x false'
-    prefix = reads[0].toString() - ~/(_R1)?(\.combined\.)?(prefixed)?(_trimmed)?(_val_1)?(\.fq)?(\.fastq)?(\.gz)?$/
+    
     fasta = "${index}/*_*.fasta"
     size = "${params.large_ref}" ? '-c' : ''
 
+    if (!params.singleEnd && skip_collapse ){
```
Comment (phue) [inline main.nf]:
```suggestion
    if (!params.singleEnd && skip_collapse){
```

Comment (phue) [reply main.nf]: whitespace

File: main.nf [hunk 5]
```diff
@@ -735,10 +801,19 @@ process bwamem {
     prefix = reads[0].toString() - ~/(_R1)?(\.combined\.)?(prefixed)?(_trimmed)?(_val_1)?(\.fq)?(\.fastq)?(\.gz)?$/
     fasta = "${index}/*.fasta"
     size = "${params.large_ref}" ? '-c' : ''
+
+    if (!params.singleEnd && skip_collapse ){
```
Comment (phue) [inline main.nf]:
```suggestion
    if (!params.singleEnd && skip_collapse){
```

Comment (phue) [reply main.nf]: whitespace

**Comment by apeltzer** [comment]
I'll be cleaning up the trimming part - its possible to have a predefined command in there for SE/PE mode and get rid of some boilerplate code this way 👍

**Comment by apeltzer** [comment]
I'll be cleaning up the trimming part - its possible to have a predefined command in there for SE/PE mode and get rid of some boilerplate code this way 👍

**Comment by apeltzer** [comment]
Unfortunately, the output handling with AdapterRemoval makes it a bit complicated to be able to accomodate all cases (PE/SE) into a single condensed version:

- depending whether you run with `--collapse`, its creating different fastq files as output
- depending on whether you use the trimming, the same happens as well

So I can't really combine individual steps in all cases, but rather have the same merging and quality trimming options in externalized commans that I use in all cases (which works fine 👍 ) to make the code a bit cleaner and easier to read. 

I guess once the tests pass, we can basically assume this is fine 👍

**Pull request merged**